### PR TITLE
Allow for combined characters in project_quick_check and Add files

### PR DESCRIPTION
**Pull request by 70ray**
This is a re-make of a similar branch but using the new unicode functions.
It shows multiple codepoints as well as bytes in the Bad Bytes tables in PQC.
Bad sequences are handled separately from invalid codepoints; this simplifies distinguishing the nature of the badness.
For Add Page it shows complete combined characters if present and title names for the bad characters.
BOM handling is also improved.

### CODE REVIEW

**Comment by 70ray** [comment]
Add files after:
image

**Comment by 70ray** [comment]
Now running in sandbox [https://www.pgdp.org/~rp31/preview](url)

File: pinc/bad_bytes.inc [hunk 1]
```diff
@@ -898,30 +898,22 @@ function get_remarks_for_bad_byte_sequence($raw)
...
-        $why_bad = "Codepoint not valid for project";
-        $intended_character = "";
-    }
     else
     {
         list($code_point, $why_bad) = $_bad_byte_sequences[$raw];
     }
 
-    if(!isset($intended_character))
+    $info = $_character_data[$code_point] ?? null;
```
Comment (cpeel) [inline pinc/bad_bytes.inc]: What you have is fine, but you might just use a ternary:
```
$char_name = isset($_character_data[$code_point]["name"]) ? $_character_data[$code_point]["name"] : '???';
```

File: pinc/project_quick_check.inc [hunk 2]
```diff
@@ -446,6 +472,17 @@ function tds_for_bad_bytes($raw)
     return $tds;
 }
 
+function tds_for_invalid_chars($raw)
+{
+    $tds = "<td>$raw</td>";
+    $tds .= "<td>" . string_to_hex($raw) . "</td>"; // bytes
+    $tds .= "<td>" . string_to_codepoints_string($raw) . "</td>"; // codepoints
+    $tds .= "<td></td>"; // intended character
+    $tds .= "<td>" . "Codepoint not valid for project" . "</td>";
```
Comment (cpeel) [inline pinc/project_quick_check.inc]: Wrap the string in `_()`.

File: pinc/bad_bytes.inc (hunk 1 above)
Comment (70ray) [reply pinc/bad_bytes.inc]: Then there could be an even shorter way:

File: pinc/project_quick_check.inc (hunk 2 above)
Comment (70ray) [reply pinc/project_quick_check.inc]: Done this.

File: pinc/bad_bytes.inc (hunk 1 above)
Comment (cpeel) [reply pinc/bad_bytes.inc]: I had to go look up the `??` operator -- very nice! Thanks for teaching me some thing new.

**Comment by srjfoo** [comment]
> This LGTM and worked in my testing. Would appreciate @srjfoo's thoughts and testing as well.

After poking around to figure out how to do combining diacritics on a Mac, and especialy multiple combining diacritics on one character -- it looks good to me, too. Screenshot:
combining_diacritics

Caveat -- you both know a whole lot more about combining diacritics than I do, so if there's something specific you want me to poke at and test, let me know. I'm just beginning to get a handle on it.

I started out by reloading various of my earlier test sets (which are simple compared to the combining diacritic issue) and everything continues to work well as far as I can see. After that, I created the new test files from which the above screenshot is taken.

**Comment by 70ray** [comment]
Now running in sandbox

[https://www.pgdp.org/~rp31/preview](url)

**Pull request merged**